core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -2933,6 +2940,112 @@ class KafkaController(val config: KafkaConfig,
+      val corruptedBroker = CorruptedBroker(brokerId, clearedFromIsrs = true)
+      zkClient.updateCorruptedBroker(corruptedBroker)
+    }
+
+    val corruptedBrokersToClearFromIsr = originalCorruptedBrokers
+      .filterNot { case (_, isrCleared) => isrCleared }.keys
+    info(s"Clearing corrupted brokers $corruptedBrokersToClearFromIsr from ISR")
+    val corruptedBrokersIsrClearResults = corruptedBrokersToClearFromIsr.map { corruptedBrokerId =>
+      val tryResult = removeBrokerIdFromPartitionsIsrs(corruptedBrokerId).flatMap { _ =>
+        updateCorruptedBrokerToCleanedInZk(corruptedBrokerId)
+      }
+      corruptedBrokerId -> tryResult
+    }
+
+    val (isrCleanSuccessful, isrCleanUnsuccessful) = corruptedBrokersIsrClearResults.partition(_._2.isSuccess)
+    // TODO: log isrCleanUnsuccessful as ERROR
+
+    val modifiedCorruptedBrokers = isrCleanSuccessful
+      .map{ case (brokerId, tryResult) => brokerId -> true }
+      .toMap
+
+    controllerContext.setCorruptedBrokers(originalCorruptedBrokers ++ modifiedCorruptedBrokers)
+  }
+
+  def removeBrokerIdFromPartitionsIsrs(brokerId: Int): Try[Unit] = Try {
+    val partitionsOnBroker = controllerContext.partitionsOnBroker(brokerId).toSeq
+    var remaining = partitionsOnBroker
+    var success = true
+    while (remaining.nonEmpty) {
+      val (finishedRemoval, removalsToRetry) = doRemoveBrokerIdFromPartitionsIsrs(brokerId, remaining)
+      remaining = removalsToRetry
+
+      finishedRemoval.foreach {
+        case (partition, Left(e)) =>
+          logger.error(s"Could not remove corrupted brokerId $brokerId from partition $partition", e)
+          success = false
+        case _ =>
+      }
+    }
+
+    if (!success) {
+      throw new KafkaException(s"Unexpected error trying to remove brokerId $brokerId from ISRs")
+    }
+  }
+
+  private def doRemoveBrokerIdFromPartitionsIsrs(brokerId: Int, partitions: Iterable[TopicPartition])
+  : (Map[TopicPartition, Either[Exception, LeaderIsrAndControllerEpoch]], Seq[TopicPartition]) = {
+    val leaderAndIsrs = partitions
+      .map(tp => tp -> controllerContext.partitionLeadershipInfo(tp))
+      .toMap
+      .filterNot { case (_, leaderAndIsrAndControllerEpochOption) => leaderAndIsrAndControllerEpochOption.isEmpty }
+      .mapValues(_.get)
+      .filter { case (_, leaderAndIsrAndControllerEpoch) => leaderAndIsrAndControllerEpoch.leaderAndIsr.isr.contains(brokerId) }
+      .mapValues(_.leaderAndIsr)
+
+    val updatedLeaderAndIsrs = leaderAndIsrs.map { case(tp, leaderAndIsr) =>
+      // Validate that brokerId is not present in ISR for any partition that has a leader
+      if (leaderAndIsr.leader != LeaderAndIsr.NoLeader) {
+        logger.warn(s"Unexpected entry $brokerId in ISR for partition $tp, leaderAndIsr = $leaderAndIsr")
+      }
+      else if (leaderAndIsr.isr.length > 1) {
+        logger.warn(s"Unexpected multiple entries in ISR for leaderless partition $tp, leaderAndIsr = $leaderAndIsr")
```

Comment:
Are these logs going to make sense to everyone else? It seems to me that they should mention "while recovering from corruption" or similar (or prefix with "Corruption-recovery: ").

Also, what's the recovery path if these occur: do nothing? (current approach) abort? manual intervention of some sort? if so, what?  In short, can this happen in any organic way, or does it imply a bug in the recovery code? You've made these WARN-level logs, but how are they actionable?

Reply:
I do not think they are actionable. They are unexpected scenarios, and should never occur ideally, so they are more like asserts. I can add a message like "This should never occur. Please investigate".

Reply:
I think just "(should never happen!)" at the end would be sufficient. They are actionable in the sense that they indicate a bug somewhere that needs to be tracked down, though not necessarily any immediate action.

Reply:
Added the message.